### PR TITLE
`indicate_inset` transform support

### PR DESCRIPTION
**Pull request by rcomer**
## PR summary

1. Corrected the `indicate_inset` docstring, as we are definitely [defaulting to transData](https://github.com/matplotlib/matplotlib/blob/84464dd085210fb57cc2419f0d4c0235391d97e6/lib/matplotlib/axes/_axes.py#L487-L488).
2. Currently the ends of the connectors that meet the box do not use the passed transform.  So against `main`, my new image test gives  
main
3. Set the `axes` property on the rectangle so that [Cartopy transforms](https://github.com/SciTools/cartopy/blob/6265416c2f05d860fe3800cbaa5d7fc6bce7d06b/lib/cartopy/crs.py#L258) will work, and we can do things like
```python
import matplotlib.pyplot as plt
import cartopy.crs as ccrs
import cartopy.feature as cfeature

main_map_proj = ccrs.TransverseMercator()
pcarree = ccrs.PlateCarree()

fig, ax = plt.subplots(subplot_kw={"projection": main_map_proj})

ax.set_extent([-50, 15, 50, 70], crs=pcarree)
ax.add_feature(cfeature.OCEAN.with_scale('50m'), facecolor="lightsteelblue", zorder=1,
               edgecolor="k", lw=0.2)

inset_ax = ax.inset_axes((0.05, 0.05, 0.3, 0.3), projection=pcarree)

inset_extent = [-44, -38, 63, 66]
inset_ax.set_extent(inset_extent, crs=pcarree)
inset_ax.add_feature(cfeature.OCEAN, facecolor="lightsteelblue", zorder=1)

ax.indicate_inset_zoom(inset_ax, edgecolor="black", transform=pcarree)

x = [inset_extent[0], inset_extent[1], inset_extent[1], inset_extent[0]]
y = [inset_extent[2], inset_extent[2], inset_extent[3], inset_extent[3]]
for a in ax, inset_ax:
    a.scatter(x, y, c=range(4), s=150, transform=pcarree)

plt.show()
```
image

## PR checklist

- [n/a] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [n/a] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [n/a] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [x] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

### CODE REVIEW

File: lib/matplotlib/inset.py [hunk 1]
```diff
@@ -171,7 +176,7 @@ def _update_connectors(self):
                 # parent artist.
                 p = ConnectionPatch(
                     xyA=xy_inset_ax, coordsA=self._inset_ax.transAxes,
-                    xyB=xy_data, coordsB=self.axes.transData,
+                    xyB=xy_data, coordsB=self._rect_trans,
```
Comment (rcomer) [inline lib/matplotlib/inset.py:179]: First I tried just using self.rectangle.get_transform() here, but an existing image test failed.  It seems the transform you get out of a rectangle is not quite what you put in?

File: lib/matplotlib/tests/test_inset.py [hunk 2]
```diff
@@ -104,3 +105,34 @@ def test_zoom_inset_connector_styles():
     # Make one visible connector a different style
     indicator.connectors[1].set_linestyle('dashed')
     indicator.connectors[1].set_color('blue')
+
+
+@image_comparison(['zoom_inset_transform.png'], remove_text=True, style='mpl20',
+                  tol=0.01)
```
Comment (rcomer) [inline lib/matplotlib/tests/test_inset.py:111]: MacOS tests need the 0.01 tolerance.

File: lib/matplotlib/inset.py (hunk 1 above)
Comment (greglucas) [reply lib/matplotlib/inset.py:179]: Do you need to use `get_patch_transform()` instead or subtract that off? There is also mention in comments of needing to add the patches to an Axes before accessing the transforms.

https://github.com/matplotlib/matplotlib/blob/84464dd085210fb57cc2419f0d4c0235391d97e6/lib/matplotlib/patches.py#L307-L309

Comment (QuLogic) [reply lib/matplotlib/inset.py:179]: > It seems the transform you get out of a rectangle is not quite what you put in?

It is not; a `Rectangle` is actually a unit square + a transform to the rectangle you really wanted, and then the user transform is added on top of that.

Comment (rcomer) [reply lib/matplotlib/inset.py]: Thanks Team Cartopy!  It looks like `get_data_transform` is what I need.

File: lib/matplotlib/inset.py [hunk 3]
```diff
@@ -79,6 +79,13 @@ def _shared_setter(self, prop, val):
 
         artist.setp([self._rectangle, *self._connectors], prop, val)
 
+    @artist.Artist.axes.setter
+    def axes(self, new_axes):
+        # Set axes on the rectangle (required for some external transforms to work) as
+        # well as the InsetIndicator artist.
+        self.rectangle.axes = new_axes
+        artist.Artist.axes.fset(self, new_axes)
```
Comment (rcomer) [inline lib/matplotlib/inset.py:87]: I copied the setter logic here from the [artist within artist example](https://matplotlib.org/stable/gallery/text_labels_and_annotations/line_with_text.html).  I would not have figured it out by myself - thanks @QuLogic for putting it in the example!

**Review by greglucas — APPROVED** [review]

**Review by QuLogic — APPROVED** [review]

**Pull request merged**